### ISSUE TITLE
交叉熵是怎么计算的？

### ISSUE THREAD

**Issue by lihe** [lihe/MyGitBlog#42]
交叉熵（**Cross Entropy**）是深度学习中最常用的 **分类损失函数**。

它衡量的是：

> **模型预测的概率分布 与 真实标签分布之间的差异。**

理解交叉熵可以从 **信息论 → 直觉 → 公式 → 二分类 → 多分类 → 为什么常用** 这几个层面来看。

------

# **一、从直觉理解交叉熵**

假设我们做 **猫狗分类**。

真实标签：

```
cat = 1
dog = 0
```

模型预测概率：

```
P(cat) = 0.9
```

说明模型预测得很好。

但如果：

```
P(cat) = 0.1
```

说明模型预测错了。

交叉熵的作用就是：

> **预测越接近真实标签，损失越小；预测越偏离真实标签，损失越大。**

------

# **二、信息论角度理解**

在信息论中：

信息量定义为：

$I(x) = -\log p(x)$

含义：

| **概率** | **信息量** |
| -------- | ---------- |
| 高概率   | 信息量小   |
| 低概率   | 信息量大   |

例如：

```
p = 0.9 → -log(0.9) ≈ 0.1
p = 0.01 → -log(0.01) ≈ 4.6
```

意思是：

如果一个模型给真实事件很低概率，就会受到很大的惩罚。

------

# **三、交叉熵公式**

假设：

- 真实分布： y
- 预测分布： p

交叉熵：

$H(y,p) = -\sum y_i \log p_i$

其中：

| **符号** | **含义** |
| -------- | -------- |
| y_i      | 真实标签 |
| p_i      | 预测概率 |

------

# **四、二分类交叉熵**

二分类最常见。

真实标签：

$y \in \{0,1\}$

预测：

$p = P(y=1)$

交叉熵：

$L = -[y\log(p) + (1-y)\log(1-p)]$

------

### **例子**

真实：

```
y = 1
```

预测：

```
p = 0.9
```

损失：

$L = -\log(0.9) ≈ 0.105$

很小。

------

如果预测：

```
p = 0.1
```

损失：

$L = -\log(0.1) ≈ 2.30$

非常大。

------

# **五、多分类交叉熵**

假设 3 类分类：

真实标签：

```
[1,0,0]
```

预测：

```
[0.7,0.2,0.1]
```

交叉熵：

$L = -\log(0.7)$

因为：

只有真实类别参与计算。

------

# **六、Softmax + Cross Entropy**

多分类模型通常：

1️⃣ 网络输出 logits：

$z = [z_1,z_2,...,z_n]$

2️⃣ Softmax 转概率：

$p_i = \frac{e^{z_i}}{\sum e^{z_j}}$

3️⃣ 计算交叉熵：

$L = -\log(p_{true})$

------

# **七、为什么交叉熵比 MSE 好**

很多人会问：

为什么不用 **均方误差（MSE）**？

原因：

### **1 梯度更好**

交叉熵梯度更稳定。

------

### **2 更符合概率模型**

？

交叉熵来自 **最大似然估计**。

------

### **3 收敛更快**

MSE 在分类任务中梯度较小。

------

# **八、交叉熵与 KL 散度**

交叉熵：

H(p,q)

可以写成：

$H(p,q) = H(p) + KL(p||q)$

其中：

| **项** | **含义**   |
| ------ | ---------- |
| H(p)   | 真实分布熵 |
| KL     | KL散度     |

因为：

H(p)

是常数。

所以最小化交叉熵 ≈ 最小化 KL 散度。

------

# **九、PyTorch 计算交叉熵**

```
import torch
import torch.nn as nn

loss = nn.CrossEntropyLoss()
```

注意：

PyTorch 会自动：

```
Softmax + CrossEntropy
```

------

# **十、一句话总结**

交叉熵的本质：

> **衡量模型预测概率分布与真实标签分布之间的差异，并对错误概率进行对数惩罚。**

公式：

二分类：

$L = -[y\log p + (1-y)\log(1-p)]$

多分类：

$L = -\sum y_i \log p_i$

------

如果你愿意，我可以再讲一个 **面试中非常常见的深入问题**：

**为什么 Softmax + CrossEntropy 的梯度会简化成**

$\hat{y} - y$

这个推导其实是 **深度学习里最经典的数学推导之一**。

**Comment by lihe** [comment]
https://zhuanlan.zhihu.com/p/438129018
